alibi_detect/utils/saving.py

```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
-    gmm_density_net = tf.keras.models.load_model(model_dir.joinpath('gmm_density_net.h5'))
-    aegmm = AEGMM(encoder_net, decoder_net, gmm_density_net, state_dict['n_gmm'], state_dict['recon_features'])
-    aegmm.load_weights(model_dir.joinpath('aegmm.ckpt'))
-    return aegmm
+    # Get backend, input_shape and detector_name
+    backend = detector.meta.get('backend', 'tensorflow')
+    if backend != 'tensorflow':
+        raise NotImplementedError("Currently, saving is only supported with backend='tensorflow'.")
+    detector_name = detector.__class__.__name__
 
+    # Process file paths
+    filepath = Path(filepath)
+    if not filepath.is_dir():
+        logger.warning('Directory {} does not exist and is now created.'.format(filepath))
+        filepath.mkdir(parents=True, exist_ok=True)
 
-def load_tf_vaegmm(filepath: Union[str, os.PathLike],
-                   state_dict: Dict) -> tf.keras.Model:
-    """
-    Load VAEGMM.
+    # Get the detector config (with artefacts still within it)
+    cfg = getattr(detector, 'get_config', None)
+    if cfg is not None:
+        cfg = cfg()  # TODO - can just do detector.get_config() once all Detectors have a .get_config()
+    else:
+        raise ValueError('{} is not supported by `save_detector`.'.format(detector_name))
 
-    Parameters
-    ----------
+    # Save x_ref
+    save_path = filepath.joinpath('x_ref.npy')
+    np.save(str(save_path), cfg['x_ref'])
+    cfg.update({'x_ref': 'x_ref.npy'})
+
+    # Save preprocess_fn
+    preprocess_fn = cfg.get('preprocess_fn', None)
+    if preprocess_fn is not None:
+        logger.info('Saving the preprocess_fn function.')
+        preprocess_cfg = _save_preprocess(preprocess_fn, backend, cfg['input_shape'], filepath)
+        cfg['preprocess_fn'] = preprocess_cfg
+
+    # Serialize kernel
+    kernel = cfg.get('kernel', None)
+    if kernel is not None:
+        device = getattr(detector, 'device', None)
+        device = device.type if device is not None else None
+        cfg['kernel'] = _save_kernel(kernel, filepath, device=device)
+        if isinstance(kernel, dict):  # serialise proj from DeepKernel
+            cfg['kernel']['proj'], _ = _save_model_config(kernel['proj'], base_path=filepath,
+                                                          input_shape=cfg['input_shape'], backend=backend)
+
+    # ClassifierDrift and SpotTheDiffDrift specific artefacts.
+    # Serialize detector model
+    model = cfg.get('model', None)
+    if model is not None:
+        model_cfg, _ = _save_model_config(model, base_path=filepath, input_shape=cfg['input_shape'], backend=backend)
+        cfg['model'] = model_cfg
+
+    # Serialize dataset
+    dataset = cfg.get('dataset', None)
+    if dataset is not None:
+        dataset_cfg, dataset_kwargs = _serialize_function(dataset, filepath, Path('dataset'))
+        cfg.update({'dataset': dataset_cfg})
+        if len(dataset_kwargs) != 0:
+            cfg['dataset']['kwargs'] = dataset_kwargs
```

Comment:
I don't think I follow! What is a `dataset` and how is it a function?

Reply:
Its a callable which is passed as a `kwarg` to some detectors e.g. [ClassiferDrift](https://docs.seldon.io/projects/alibi-detect/en/latest/cd/methods/classifierdrift.html). Usually, it would be a [TFDataset](https://docs.seldon.io/projects/alibi-detect/en/latest/api/alibi_detect.utils.tensorflow.data.html?highlight=TFdataset#alibi_detect.utils.tensorflow.data.TFDataset) or the pytorch equivalent. Since it is pre-registered, `_serialize_function` will save it to the `config.toml` as a function registry (i.e. an `'@...'` string). 

We should probably open an issue to document `TFDataset` etc better, as it could currently be confused with this page:
https://docs.seldon.io/projects/alibi-detect/en/latest/datasets/overview.html

Reply:
Correction: `dataset` is not a model. `TFDataset` is a [tf.keras.utils.Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence), which can be pickled. Hence we are fine here.